sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala

```diff
@@ -155,3 +166,160 @@ case class GpuShuffledHashJoinExec(
+   * to also pull to host its first batch. After the first stream batch is on the host, the
+   * stream iterator acquires the semaphore and then the build side is copied to the GPU.
+   *
+   * Prior to this we would get a build batch on the GPU, acquiring
+   * the semaphore in the process, and then begin pulling from the stream iterator,
+   * which could include IO (while holding onto the semaphore).
+   *
+   * The function handles the case where the build side goes above the configured batch
+   * goal, in which case it will concat on the host, grab the semaphore, and continue to
+   * pull the build iterator to build a bigger batch on the GPU. This is not optimized
+   * because we hold onto the semaphore during the entire time after realizing the goal
+   * has been hit.
+   *
+   * @param hostTargetBatchSize target batch size goal on the host
+   * @param buildOutput output attributes of the build plan
+   * @param buildIter build iterator
+   * @param streamIter stream iterator
+   * @param spillCallback metric updater in case downstream iterators spill
+   * @param coalesceMetricsMap metrics map with metrics to be used in downstream
+   *                           iterators
+   * @return a pair of `ColumnarBatch` and streamed iterator that can be
+   *         used for the join
+   */
+  def getBuiltBatchAndStreamIter(
+      hostTargetBatchSize: Long,
+      buildOutput: Seq[Attribute],
+      buildIter: Iterator[ColumnarBatch],
+      streamIter: Iterator[ColumnarBatch],
+      spillCallback: SpillCallback,
+      coalesceMetricsMap: Map[String, GpuMetric]): (ColumnarBatch, Iterator[ColumnarBatch]) = {
+    val semWait = coalesceMetricsMap(GpuMetric.SEMAPHORE_WAIT_TIME)
+    val buildTime = coalesceMetricsMap(GpuMetric.BUILD_TIME)
+    var bufferedBuildIterator: CloseableBufferedIterator[ColumnarBatch] = null
+    closeOnExcept(bufferedBuildIterator) { _ =>
+      val startTime = System.nanoTime()
+      // find if the build side is non-empty, and if the first batch is
+      // a serialized batch. If neither condition is met, we fallback to the
+      // `getSingleBatchWithVerification` method.
+      val firstBatchIsSerialized = {
+        if (!buildIter.hasNext) {
+          false
+        } else {
+          bufferedBuildIterator = new CloseableBufferedIterator(buildIter.buffered)
+          val firstBatch = bufferedBuildIterator.head
+          if (firstBatch.numCols() != 1) {
+            false
+          } else {
+            firstBatch.column(0) match {
+              case _: SerializedTableColumn => true
+              case _ => false
+            }
+          }
+        }
+      }
+
+      if (!firstBatchIsSerialized) {
+        // fallback if we failed to find serialized build batches
+        val builtBatch =
+          ConcatAndConsumeAll.getSingleBatchWithVerification(
+            Option(bufferedBuildIterator).getOrElse(buildIter), buildOutput)
```

Comment:
There should be no need to emulate a fallback here.  If the first batch isn't serialized then that means we're not a join that was directly connected to a shuffle on the build-side.  That means the coalesce should have already occurred, otherwise there's something wrong with the planning.  May be prudent to `assert(!buildIter.hasNext)` but otherwise we should be able to just use the build batch and stream iter directly at this point.

Reply:
OK, sounds like a plan

Reply:
Thinking about this one more, I am not seeing the issue, I am probably missing something. The call to `getSingleBatchWithVerification` is building an empty batch (if the build itereator is empty) and is checking that it gets a single GPU batch in the non-empty build iterator case.